### DISCUSSION TITLE
Set helpers on a custom generated token set?

### DISCUSSION THREAD

**Discussion by m1212e** [panva/openid-client#709] [Q&A] [answered]
Before the major release I could init a `new TokenSet({ ... a token set parsed from a cookie ... })` and then run helpers like

```ts
if (tokenSet.expired() && tokenSet.data.refresh_token) {
		try {
			tokenSet = await refresh(tokenSet.data.refresh_token);
```
on that instance. I see how I can refresh a token set via `refreshTokenGrant` but can I also somehow check the expiry without implementing that logic myself?

Basically I want access to this:
https://github.com/panva/openid-client/blob/ff1436ea991d771b0257a7c12752621ae6abaaf9/src/index.ts#L1747

**Comment by panva** [comment, marked as answer]
Those helpers depended on rewriting the Token Endpoint Response from having `expires_in` to having an `expires_at` and then `expires_in` being a calculated getter. And the return was it's own Class 🙅.

I don't want to do that anymore, the intention is to merely return what the AS returned. To that end I suggest that you calculate the expiration time yourself based on the expires_in _when_ you receive it.

There's an interaction with missing `expires_in` that I simply don't want to deal with anymore. The AS doesn't have to return it, not to mention that the AS may revoke the Access Token at any point in time anyway.

Alternatively, just use the token until you get `WWWAuthenticateChallengeError` and then refresh.

> Basically I want access to this:

Yeah that wouldn't work because it depends on knowing when the token endpoint response was received from the AS.

**Reply by m1212e** [reply]
I see, thanks for the quick response!
